gossip/mempool_test.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func Test_BloomFilterAdd(t *testing.T) {
+	tests := []struct {
+		name     string
+		add      [][]byte
+		expected [][]byte
+	}{
+		{
+			name:     "add nothing",
+			add:      [][]byte{},
+			expected: [][]byte{},
+		},
+		{
+			name: "add single element",
+			add: [][]byte{
+				[]byte("00000000"),
+			},
+			expected: [][]byte{
+				[]byte("00000000"),
+			},
+		},
+		{
+			name: "add multiple elements",
+			add: [][]byte{
+				[]byte("00000000"),
+				[]byte("00000001"),
+				[]byte("00000002"),
+				[]byte("00000003"),
+				[]byte("00000004"),
+			},
+			expected: [][]byte{
+				[]byte("00000000"),
+				[]byte("00000001"),
+				[]byte("00000002"),
+				[]byte("00000003"),
+				[]byte("00000004"),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := require.New(t)
+			b, err := newBloomFilter(10, 1)
+			r.NoError(err)
+
+			for _, item := range tt.add {
+				b.Add(item)
+			}
+
+			r.Equal(b.bloomFilter.N(), uint64(len(tt.expected)))
+
+			for _, expected := range tt.expected {
+				r.True(b.Contains(expected))
+			}
+		})
+	}
+}
+
+func Test_BloomFilterAdd_Refresh(t *testing.T) {
+	tests := []struct {
+		name         string
+		refreshRatio float64
+		add          [][]byte
+		expected     [][]byte
+	}{
+		{
+			name:         "no refresh",
+			refreshRatio: 1,
+			add: [][]byte{
+				[]byte("00000000"),
+			},
+			expected: [][]byte{
+				[]byte("00000000"),
+			},
+		},
+		{
+			name:         "refresh",
+			refreshRatio: 0.1,
+			add: [][]byte{
+				[]byte("00000000"),
+				[]byte("00000001"),
+			},
+			expected: [][]byte{
+				[]byte("00000001"),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := require.New(t)
+			b, err := newBloomFilter(10, 1)
+			r.NoError(err)
+
+			for _, item := range tt.add {
+				_ = ResetBloomFilterIfNeeded(b, tt.refreshRatio)
+				b.Add(item)
+			}
+
+			r.Equal(b.bloomFilter.N(), uint64(len(tt.expected)))
+
+			for _, expected := range tt.expected {
+				r.True(b.Contains(expected))
+			}
+		})
+	}
+}
+
+func TestBloomFilter_MarshalBinary(t *testing.T) {
+	r := require.New(t)
+
+	original, err := NewBloomFilter()
+	r.NoError(err)
+
+	bytes, err := original.MarshalBinary()
+	r.NoError(err)
+
+	unmarshalled, err := NewBloomFilter()
+	r.NoError(err)
+	r.NoError(unmarshalled.UnmarshalBinary(bytes))
+
+	r.Equal(original, unmarshalled)
+}
```

Comment:
could we change this to take in a short set of test cases?

Also, what would we need to break in the code to make this test fail? It seems like this is really testing the bloom filter library that we're using. I think this is good to have either way, so just thinking out loud based off a story @richardpringle told me a few weeks ago

Reply:
Removed these test cases